jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp

```diff
@@ -0,0 +1,369 @@
+  // Mapping from global variables declared with delta nodes to PointerObject
+  std::unordered_map<const delta::node *, PointerObject::Index> GlobalMap_;
+  // Mapping from functions declared with lambda nodes to PointerObject
+  std::unordered_map<const lambda::node *, PointerObject::Index> FunctionMap_;
+  // Mapping from symbols imported into the module to PointerObject
+  std::unordered_map<const rvsdg::argument *, PointerObject::Index> ImportMap_;
+
+  /**
+   * Internal helper function for adding PointerObjects, use the Create* methods instead
+   */
+  PointerObject::Index
+  AddPointerObject(PointerObjectKind kind);
+
+public:
+
+  /**
+   * Creates a PointerObject of register kind, and maps the rvsdg output to the new PointerObject.
+   * @param rvsdgOutput the rvsdg output associated with the register PointerObject
+   * @return the index of the new PointerObject, in the PointerObjectSet
+   */
+  PointerObject::Index
+  CreateRegisterPointerObject(const rvsdg::output &rvsdgOutput);
+
+  /**
+   * Reuses an existing PointerObject of register type for an additional rvsdg output.
+   * This is useful when two rvsdg outputs can be shown to always hold the exact same value.
+   * @param rvsdgOutput the new rvsdg output providing the register value
+   * @param pointerObject the index of the existing PointerObject, must be of register kind
+   */
+  void
+  MapRegisterToExistingPointerObject(const rvsdg::output &rvsdgOutput, PointerObject::Index pointerObject);
+
+  PointerObject::Index
+  CreateAllocaMemoryObject(const rvsdg::node &allocaNode);
+
+  PointerObject::Index
+  CreateMallocMemoryObject(const rvsdg::node &mallocNode);
+
+  PointerObject::Index
+  CreateGlobalMemoryObject(const delta::node &deltaNode);
+
+  PointerObject::Index
+  CreateFunctionMemoryObject(const lambda::node &lambdaNode);
+
+  PointerObject::Index
+  CreateImportMemoryObject(const rvsdg::argument &importNode);
+
+  const std::unordered_map<const rvsdg::output *, PointerObject::Index>&
+  GetRegisterMap() const noexcept;
+
+  const std::unordered_map<const rvsdg::node *, PointerObject::Index>&
+  GetAllocaMap() const noexcept;
+
+  const std::unordered_map<const rvsdg::node *, PointerObject::Index>&
+  GetMallocMap() const noexcept;
+
+  const std::unordered_map<const delta::node *, PointerObject::Index>&
+  GetGlobalMap() const noexcept;
+
+  const std::unordered_map<const lambda::node *, PointerObject::Index>&
+  GetFunctionMap() const noexcept;
+
+  const std::unordered_map<const rvsdg::argument *, PointerObject::Index>&
+  GetImportMap() const noexcept;
+
+  PointerObject::Index
+  NumPointerObjects() const noexcept;
+
+  PointerObject &
+  GetPointerObject(PointerObject::Index index);
+
+  const PointerObject &
+  GetPointerObject(PointerObject::Index index) const;
+
+  const std::unordered_set<PointerObject::Index> &
+  GetPointsToSet(PointerObject::Index idx) const;
+
+  /**
+   * Adds pointee to P(pointer)
```

Comment:
You have nowhere defined what `P(pointer)` is. Maybe: "Adds \p pointee to pointer object set of \p pointer"